solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformerFactory.java

```diff
@@ -70,109 +73,59 @@ public DocTransformer create(String field, SolrParams params, SolrQueryRequest r
-    return name;
-  }
-  
-  @Override
-  public String[] getExtraRequestFields() {
-    // we always need the idField (of the parent) in order to fill out it's children
-    return new String[] { idField.getName() };
+    if(buildHierarchy) {
+      if(childFilter != null) {
+        childFilter = buildHierarchyChildFilterString(childFilter);
+        return new ChildDocTransformer(field, parentsFilter, req,
+            getChildQuery(childFilter, req), limit);
+      }
+      return new ChildDocTransformer(field, parentsFilter, req, null, limit);
+    }
+    return new ChildDocTransformer( field, parentsFilter, req,
+        childFilter==null? null: getChildQuery(childFilter, req), limit);
   }
 
-  @Override
-  public void transform(SolrDocument doc, int docid) {
-
-    FieldType idFt = idField.getType();
-    Object parentIdField = doc.getFirstValue(idField.getName());
-    
-    String parentIdExt = parentIdField instanceof IndexableField
-      ? idFt.toExternal((IndexableField)parentIdField)
-      : parentIdField.toString();
-
+  private static Query getChildQuery(String childFilter, SolrQueryRequest req) {
     try {
-      Query parentQuery = idFt.getFieldQuery(null, idField, parentIdExt);
-      Query query = new ToChildBlockJoinQuery(parentQuery, parentsFilter);
-      DocList children = context.getSearcher().getDocList(query, childFilterQuery, new Sort(), 0, limit);
-      if(children.matches() > 0) {
-        SolrDocumentFetcher docFetcher = context.getSearcher().getDocFetcher();
-
-        Set<String> dvFieldsToReturn = docFetcher.getNonStoredDVs(true);
-        boolean shouldDecorateWithDVs = dvFieldsToReturn.size() > 0;
-        DocIterator i = children.iterator();
-
-        while(i.hasNext()) {
-          Integer childDocNum = i.next();
-          Document childDoc = context.getSearcher().doc(childDocNum);
-          // TODO: future enhancement...
-          // support an fl local param in the transformer, which is used to build
-          // a private ReturnFields instance that we use to prune unwanted field 
-          // names from solrChildDoc
-          SolrDocument solrChildDoc = DocsStreamer.convertLuceneDocToSolrDoc(childDoc, schema,
-                                                                             new SolrReturnFields());
+      return QParser.getParser( childFilter, req).getQuery();
+    } catch (SyntaxError syntaxError) {
+      throw new SolrException( ErrorCode.BAD_REQUEST, "Failed to create correct child filter query" );
+    }
+  }
 
-          if (shouldDecorateWithDVs) {
-            docFetcher.decorateDocValueFields(solrChildDoc, childDocNum, dvFieldsToReturn);
-          }
-          doc.addChildDocument(solrChildDoc);
-        }
-      }
-      
-    } catch (IOException e) {
-      doc.put(name, "Could not fetch child Documents");
+  protected static String buildHierarchyChildFilterString(String queryString) {
+    List<String> split = StrUtils.splitSmart(queryString, ':');
```

Comment:
FYI I've rewritten this method to not use splitting & joining in cases like this which can use "indexOf"/"lastIndexOf".   I've also increased it's robustness to more complicated query examples of multiple conditions.  

For now I think we shouldn't document this; let it be kinda a secret feature until we can query (in q/fq) in like-kind.